### PR TITLE
Added add_default_join plugin hook

### PR DESCRIPTION
**Pull request by edgardmessias**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none

### CODE REVIEW

File: inc/search.class.php [hunk 1]
```diff
@@ -4713,43 +4713,42 @@ static function addWhere($link, $nott, $itemtype, $ID, $searchtype, $val, $meta
...
          case 'User' :
-            return self::addLeftJoin($itemtype, $ref_table, $already_link_tables,
+            $out = self::addLeftJoin($itemtype, $ref_table, $already_link_tables,
                                      "glpi_profiles_users", "profiles_users_id", 0, 0,
                                      ['jointype' => 'child']);
+            break;
 
          case 'Reminder' :
-            return Reminder::addVisibilityJoins();
+            $out = Reminder::addVisibilityJoins();
```
Comment (cedric-anne) [inline inc/search.class.php:4727]:
```suggestion
            $out = Reminder::addVisibilityJoins();
            break;
```

File: inc/search.class.php [hunk 2]
```diff
@@ -4713,43 +4713,42 @@ static function addWhere($link, $nott, $itemtype, $ID, $searchtype, $val, $meta
...
 
          case 'RSSFeed' :
-            return RSSFeed::addVisibilityJoins();
+            $out = RSSFeed::addVisibilityJoins();
```
Comment (cedric-anne) [inline inc/search.class.php:4730]:
```suggestion
            $out = RSSFeed::addVisibilityJoins();
            break;
```

File: inc/search.class.php [hunk 3]
```diff
@@ -4713,43 +4713,42 @@ static function addWhere($link, $nott, $itemtype, $ID, $searchtype, $val, $meta
...
          case 'ProjectTask' :
             // Same structure in addDefaultWhere
-            $out  = '';
             $out .= self::addLeftJoin($itemtype, $ref_table, $already_link_tables,
                                       "glpi_projects", "projects_id");
             $out .= self::addLeftJoin($itemtype, $ref_table, $already_link_tables,
                                       "glpi_projecttaskteams", "projecttaskteams_id", 0, 0,
                                       ['jointype' => 'child']);
-            return $out;
+                                      break;
```
Comment (cedric-anne) [inline inc/search.class.php:4739]:
```suggestion
            break;
```

File: inc/search.class.php [hunk 4]
```diff
@@ -4871,13 +4870,13 @@ static function addDefaultJoin($itemtype, $ref_table, array &$already_link_table
...
-               if (!empty($out)) {
-                  return $out;
-               }
             }
-
-            return "";
+            break;
       }
+
+      /* Hook to restrict user right on current itemtype */
```
Comment (cedric-anne) [inline inc/search.class.php:4877]:
```suggestion
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]
Can you explain how you want to use this hook, and maybe give an example of code ?

**Comment by cedric-anne** [comment]
I changed target branch to master, as 9.5/bugfixes branch is only for bug fixes.

**Comment by edgardmessias** [comment]
> Can you explain how you want to use this hook, and maybe give an example of code ?

Sometimes when I am developing a internal plugin, I see the necessity of change the default  filter behavior, and using only `add_default_where` it's a performance nightmare

**Review by cedric-anne — APPROVED** [review]

**Comment by cedric-anne** [comment]
Hi,

Can you add this in documentation (https://github.com/glpi-project/docdev/blob/master/source/plugins/hooks.rst) and also add a short example of usage in example plugin (https://github.com/pluginsGLPI/example) ?

Regards

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Pull request merged**